sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1553,13 +1553,84 @@ object ReplaceDeduplicateWithAggregate extends Rule[LogicalPlan] {
  * 1. This rule is only applicable to INTERSECT DISTINCT. Do not use it for INTERSECT ALL.
  * 2. This rule has to be done after de-duplicating the attributes; otherwise, the generated
  *    join conditions will be incorrect.
+ *
+ * This rule also pushed down the Distinct to left/right side based on the config
+ *   `SQLConf.OPTIMIZE_INTERSECT_ENABLED`
+ * 1. It pushes down Distinct on left side if applying a Distinct on left reduces rows by
+ *   a configured threshold
+ * 2. It pushes down Distinct on right side if applying a Distinct on right side converts the join
+ *   type to broadcast join, or it reduces rows by a configured threshold
+ * If Distinct is pushdown on left side, it won't apply Distinct after Join as it is not needed.
  */
 object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
+  private def canBroadcast(plan: LogicalPlan): Boolean = {
+    plan.stats.sizeInBytes >= 0 && plan.stats.sizeInBytes <= SQLConf.get.autoBroadcastJoinThreshold
+  }
+
+  private def shouldApplyDistinctToRightSide(right: LogicalPlan): Boolean = {
+    if (!SQLConf.get.optimizeIntersectEnabled) {
+      false
+    } else if (!canBroadcast(right) && canBroadcast(Distinct(right))) {
```

Comment:
Should we check only for `canBroadcast(Distinct(right)`, since `canBroadcast(right)` will be checked during physical planning in `JoinSelection`.. That way we could avoid having to copy `canBroadcast` logic?